### ISSUE TITLE
Finally blocks need to be executed

### ISSUE THREAD

**Issue by kelunik** [fiberphp/fiber-ext#26]
When a fiber is unfinished and GCed, it must execute all finally blocks.

**Comment by uasan** [comment]
Is `return` it also supposed to work?

```PHP
new Fiber(function () {
    try {
        echo "yielding - "
        
        return Fiber::yield();
    }
    finally {
        echo "finally";
    }
});
```

**Comment by kelunik** [comment]
Yes, same there, compare with https://3v4l.org/Zq0BZ.

**Comment by lvht** [comment]
It seems impossible to implement this feature without modify the zend vm. @kelunik

**Comment by kelunik** [comment]
@lvht It's probably time to start the work on a php-src PR then? @bwoebi already started looking into moving internal functions into the VM, so things like `array_map` work.

**Comment by lvht** [comment]
Could you @bwoebi offer detail about your solution?

**Comment by bwoebi** [comment]
One can, theoretically, integrate it separately from the zend VM:

It's possible to unconditionally hook into the zend VM by throwing a fake exception (i.e. EG(exception) != NULL). One can at the same time override the exception handler (EG(exception_op)[0].handler). At that place you can (after restoring the exception handler) easily load your own execute_data and call ZEND_VM_ENTER(); - now you just have to rewrite these functions currently using zend_call_function so that they save data onto the VM stack (just set EG(vm_stack_top) to a value high enough for that data) and read from it when a callback is called. That callback is called with the return value from your custom opcode handler each time a called userland (or internal function) returns.

The thing I have to still look into is exception handling, when your callback is throwing. Afterwards I'll upload my PoC (it's integrated into the zend VM but should be easy to separate out).

**Comment by kelunik** [comment]
According to the tests @martinschroeder managed to solve this.

**Comment by martinschroeder** [comment]
Yes, finally blocks are executed because i am throwing an error into the fiber when it is suspended while being garbage collected. The free object handler will set the fibers status to DEAD and resume the fiber by throwing an error into it at the position of the current yield. Any further calls to Fiber::yield() will mmediately throw another error, preventing the fiber from being suspended again.

Since these are standard PHP errors they can be handles using try / catch. When a garbage collected fiber execution terminates with an error that error will silently be ignored. Care must be taken not to catch errors in an infinite loop in PHP code without exiting the loop, because this will be similar to calling while (1); which shoul obviously be avoided. ;-)
